optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/uni/DroolsGroupByInvoker.java

```diff
@@ -0,0 +1,101 @@
+import org.drools.core.spi.Accumulator;
+import org.drools.core.spi.CompiledInvoker;
+import org.drools.core.spi.Tuple;
+import org.optaplanner.core.api.score.stream.uni.UniConstraintCollector;
+
+public class DroolsGroupByInvoker<A, B, ResultContainer, NewB> implements Accumulator, CompiledInvoker {
+
+    private final UniConstraintCollector<B, ResultContainer, NewB> collector;
+
+    public DroolsGroupByInvoker(final UniConstraintCollector<B, ResultContainer, NewB> collector) {
+        this.collector = collector;
+    }
+
+    @Override
+    public Serializable createContext() {
+        return new DroolsGroupBy<>(collector);
+    }
+
+    @Override
+    public void init(Object workingMemoryContext, Object context, Tuple tuple, Declaration[] declarations,
+            WorkingMemory workingMemory) {
+        ((DroolsGroupBy<A, B, ResultContainer, NewB>) context).init();
+    }
+
+    @Override
+    public void accumulate(Object workingMemoryContext, Object context, Tuple tuple, InternalFactHandle handle,
+            Declaration[] declarations, Declaration[] innerDeclarations, final WorkingMemory workingMemory) {
+        InternalWorkingMemory internalWorkingMemory = (InternalWorkingMemory) workingMemory;
+        Object handleObject = handle.getObject();
+        final B toCollect = (B) innerDeclarations[1].getValue(internalWorkingMemory, handleObject);
+        final A groupKey = (A) innerDeclarations[2].getValue(internalWorkingMemory, handleObject);
+        ((DroolsGroupBy<A, B, ResultContainer, NewB>) context).accumulate(handle, groupKey, toCollect);
+    }
+
+    @Override
+    public void reverse(Object workingMemoryContext, Object context, Tuple tuple, InternalFactHandle handle,
+            Declaration[] declarations, Declaration[] innerDeclarations, WorkingMemory workingMemory) {
+        ((DroolsGroupBy<A, B, ResultContainer, NewB>) context).reverse(handle);
+    }
+
+    @Override
+    public Object getResult(Object workingMemoryContext, Object context, Tuple tuple, Declaration[] declarations,
+            WorkingMemory workingMemory) {
+        return ((DroolsGroupBy<A, B, ResultContainer, NewB>) context).getResult();
+    }
+
+    @Override
+    public boolean supportsReverse() {
+        return true;
+    }
+
+    @Override
+    public Object createWorkingMemoryContext() {
+        return null;
+    }
+
+    @Override
+    public String getMethodBytecode() {
+        Class<?> accumulateClass = DroolsGroupBy.class;
+        String classFileName = accumulateClass.getCanonicalName().replace('.', '/') + ".class";
+        try (InputStream is = getClass().getClassLoader().getResourceAsStream(classFileName)) {
+            final byte[] data = new byte[1024];
+            int byteCount;
+            ByteArrayOutputStream bos = new ByteArrayOutputStream();
+            while ((byteCount = is.read(data, 0, 1024)) > -1) {
+                bos.write(data, 0, byteCount);
+            }
+            return bos.toString();
```

Comment:
This `toString()` is platform-dependent; maybe we want to pass a charset there.

Reply:
After talking to Radovan, we think this is not an issue. We're reading a classfile, and classfile encoding must be platform independent. (Otherwise it would defeat the purpose of bytecode, to be the same on all platforms.)